File: emp_ui/lib/models/salary.dart
class SalaryModel {
  final int? id;
  final int? employeeId;
  final String salaryCode;
  final String name;
  final String phone;
  final double basicSalary;
  final double houseRent;
  final double medical;
  final double transport;
  final double overTimeRate;
  final double overTimeHour;
  final double grossSalary;
  final String bankName;

  SalaryModel({
    this.id,
    this.employeeId,
    required this.salaryCode,
    required this.name,
    required this.phone,
    required this.basicSalary,
    required this.houseRent,
    required this.medical,
    required this.transport,
    required this.overTimeRate,
    required this.overTimeHour,
    required this.grossSalary,
    required this.bankName,
  });

  factory SalaryModel.fromJson(Map<String, dynamic> json) {
    return SalaryModel(
      id: json['id'],
      employeeId: json['employee'] != null ? json['employee']['id'] : (json['employeeId'] ?? json['employee_id']),
      salaryCode: json['salaryCode'] ?? '',
      name: json['name'] ?? '',
      phone: json['phone'] ?? '',
      basicSalary: (json['basicSalary'] as num?)?.toDouble() ?? 0.0,
      houseRent: (json['houseRent'] as num?)?.toDouble() ?? 0.0,
      medical: (json['medical'] as num?)?.toDouble() ?? 0.0,
      transport: (json['transport'] as num?)?.toDouble() ?? 0.0,
      overTimeRate: (json['overTimeRate'] as num?)?.toDouble() ?? 0.0,
      overTimeHour: (json['overTimeHour'] as num?)?.toDouble() ?? 0.0,
      grossSalary: (json['grossSalary'] as num?)?.toDouble() ?? 0.0,
      bankName: json['bankName'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {
      "salaryCode": salaryCode,
      "name": name,
      "phone": phone,
      "basicSalary": basicSalary,
      "houseRent": houseRent,
      "medical": medical,
      "transport": transport,
      "overTimeRate": overTimeRate,
      "overTimeHour": overTimeHour,
      "grossSalary": grossSalary,
      "bankName": bankName,
    
      "employeeId": employeeId,
      "employee_id": employeeId,
      "employee": {"id": employeeId},
    };

    if (id != null) {
      data["id"] = id;
    }

    return data;
  }
}

File: emp-api/src/main/java/com/example/entities/Salary.java
package com.example.entities;

import jakarta.persistence.*;

@Entity
@Table(name = "salary_sheet")
public class Salary {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;
    private Integer employeeId;
    private String salaryCode;
    private String name;
    private String phone;

    private double basicSalary;
    private double houseRent;
    private double medical;
    private double transport;

    private double overTimeRate;
    private double overTimeHour;

    private double grossSalary;

    private String bankName;

    // ================= Calculation =================

    private void calculateSalary() {
        this.houseRent = basicSalary * 0.20;
        this.medical = basicSalary * 0.10;
        this.transport = basicSalary * 0.15;

        double overtimeAmount = overTimeRate * overTimeHour;

        this.grossSalary = basicSalary
                + houseRent
                + medical
                + transport
                + overtimeAmount;
    }

    // ================= Getter & Setter =================
    public Integer getEmployeeId() {
        return employeeId;
    }

    public void setEmployeeId(Integer employeeId) {
        this.employeeId = employeeId;
    }
    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getSalaryCode() {
        return salaryCode;
    }

    public void setSalaryCode(String salaryCode) {
        this.salaryCode = salaryCode;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getPhone() {
        return phone;
    }

    public void setPhone(String phone) {
        this.phone = phone;
    }

    public double getBasicSalary() {
        return basicSalary;
    }

    public void setBasicSalary(double basicSalary) {
        this.basicSalary = basicSalary;
        calculateSalary();
    }

    public double getHouseRent() {
        return houseRent;
    }

    public double getMedical() {
        return medical;
    }

    public double getTransport() {
        return transport;
    }

    public double getOverTimeRate() {
        return overTimeRate;
    }

    public void setOverTimeRate(double overTimeRate) {
        this.overTimeRate = overTimeRate;
        calculateSalary();
    }

    public double getOverTimeHour() {
        return overTimeHour;
    }

    public void setOverTimeHour(double overTimeHour) {
        this.overTimeHour = overTimeHour;
        calculateSalary();
    }

    public double getGrossSalary() {
        return grossSalary;
    }

    public String getBankName() {
        return bankName;
    }

    public void setBankName(String bankName) {
        this.bankName = bankName;
    }
}

